#!/usr/bin/env Rscript
## Annotate the results of gene-based (rare variant) association tests as implemented in SKAT package using PLINK format files
## Written by Fahri Kucukali for the analyses conducted for Kucukali et al., Whole‐exome rare‐variant analysis of Alzheimer's disease and related biomarker traits, Alzheimer's & Dementia (2022). http://doi.org/10.1002/alz.12842
## For issues: https://github.com/SleegersLab-VIBCMN/AD_Biomarkers_RareVariantAnalyses or e-mail Fahri.Kucukali@uantwerpen.vib.be

suppressWarnings(suppressMessages(library(SKAT)))
suppressWarnings(suppressMessages(library(argparser)))
suppressWarnings(suppressMessages(library(dplyr)))
suppressWarnings(suppressMessages(library(data.table)))
suppressWarnings(suppressMessages(library(ggplot2)))
suppressWarnings(suppressMessages(library(ggrepel)))
suppressWarnings(suppressMessages(library(ggbeeswarm)))
suppressWarnings(suppressMessages(library(scales)))
suppressWarnings(suppressMessages(library(qqman)))

p <- arg_parser("Annotate PLINKSKAT.R or any SKAT/SKAT-O results, with different modes and options. Publication Version 1.1, December 2022")
p <- add_argument(p, "--mode", help = "Type full or oneset. Full annotates all results, oneset focus on one gene indicated by --gene.", type = "character", nargs = '+')
p <- add_argument(p, "--gene", help = "indicate the gene name you want to test, only valid if --mode oneset is selected. Optional.", type="character", nargs = '*')
p <- add_argument(p, "--variable", help = "Enter binary or continuous.", type = "character", nargs = '+')
p <- add_argument(p, "--result", help = "Results file name from PLINKSKAT.R or any SKAT/SKAT-O tests.", nargs = '+')
p <- add_argument(p, "--info", help = ".info file name generated by PLINKSKAT.R (dummy argument) or any SKAT/SKAT-O tests.", nargs = '+')
p <- add_argument(p, "--ssd", help = ".ssd file name generated by PLINKSKAT.R (dummy argument) or any SKAT/SKAT-O tests.", nargs = '+')
p <- add_argument(p, "--phenotype", help = "Tested phenotype name.", type = "character", nargs = '+')
p <- add_argument(p, "--phenocovar", help = "Phenocovar file name.", type = "character", nargs = '+')
p <- add_argument(p, "--covnames", help = "Covariates you used to perform PLINKSKAT.R or any SKAT/SKAT-O tests. Optional.", nargs = '*')
p <- add_argument(p, "--geneAnnotFile", help = "An annotation file for tested genes, with header, first column is gene names as supplied in SetID. Rest of the columns can be any information, but chr (without chr prefix) and start columns are required to generate Manhattan plots.", type = "character", nargs = '+')
p <- add_argument(p, "--plotAndTablePath", help = "Paths of the plots and tables that will be created for a specific threshold.",type="character", nargs = '+')
p <- add_argument(p, "--threshold", help = "Thresholds for plots and tables created for full analysis, or the given gene SetID; also used for labelling the manhattan plots.", type="numeric", nargs = '+')
p <- add_argument(p, "--fam", help = "Plink FAM file, to make sure phenocovar and fam files are sorted the same way.", type = "character", nargs = '+')
p <- add_argument(p, "--missing_cutoff", help = "Defines the call rate cutoff for variants, default is 0.15 meaning that keeping all variants present in at least 85% of individuals.", nargs = 1, type = "numeric", default = 0.15)
p <- add_argument(p, "--carrier_cutoff", help = "Defines the carrier cutoff, default is 2 meaning that genes that did not have at least 2 carriers for its tested variants in analyses will be removed.", nargs = 1, type = "numeric", default = 2)
p <- add_argument(p, "--testtype", help = "Only for annotating plots, enter whether you tested protein-altering or LoF variants. Optional.", type = "character", nargs = '*', default = "")

arg <- parse_args(p)

# argument checks
if (is.na(arg$mode) || is.na(arg$variable) || is.na(arg$result) || is.na(arg$info) || is.na(arg$ssd) || is.na(arg$phenotype) || is.na(arg$phenocovar) || is.na(arg$geneAnnotFile) || is.na(arg$plotAndTablePath) || is.na(arg$threshold) || is.na(arg$fam)) {
    message("Error: exiting because at least one of the following required arguments is missing: --mode, --variable, --result, --info, --ssd , --phenotype, --phenocovar, --geneAnnotFile, --plotAndTablePath, --threshold, --fam")
    q("no")
}

if ((arg$mode == "oneset") && is.na(arg$gene)) {
    message("Error: exiting because --gene argument is missing, even though mode is set as oneset. Please provide it when you use --mode oneset.")
    q("no")
}

ssd <- arg$ssd
info <- arg$info

phenotype <- paste0(arg$phenotype,"")

results <- fread(arg$result, header=TRUE)

resultname <- paste0(arg$result,"")

plotAndTablePath <- paste0(arg$plotAndTablePath,"")

geneAnnotFile <- fread(arg$geneAnnotFile, header=TRUE)

colnames(geneAnnotFile)[which(names(geneAnnotFile) == "gene_name")] <- "SetID"

phenocovar <- fread(arg$phenocovar, header=TRUE)

fam_sort_order <- fread(arg$fam, header=FALSE, select="V2")

sortingorder <- match(fam_sort_order$V2, phenocovar$IID)

phenocovar <- phenocovar[sortingorder,]

no_cols <- ncol(phenocovar) + 1

ssd_info <- Open_SSD(ssd,info)

if(arg$mode == "full") {

  if (arg$variable == "binary") {

	output_prepare <- as.data.frame(matrix(0, 1, 13))

	colnames(output_prepare) <- c("SetID","cases_cMAC","cases_cMAF","controls_cMAC","controls_cMAF", "no_of_carriers", "enrichment_ratio", "P_regression", "OR_noOfMutations","OR_L95","OR_U95","eff_estimate","eff_stdErr")

	for (i in results$SetID){

	SetID <- i

	set_index <- ssd_info$SetInfo[ssd_info$SetInfo$SetID == i,]$SetIndex

	table <- cbind(phenocovar,as.data.frame(Get_Genotypes_SSD(ssd_info, set_index, is_ID = TRUE)))

	# Convert 9s in genotypes to NAs, and drop variants depending on the missing_cutoff within this subset to be in corcordant

	table[,no_cols:ncol(table)] <- table[,no_cols:ncol(table)] %>% mutate_all(~na_if(., 9))

	call_rate <- 1 - arg$missing_cutoff

	table <- data.table(data.frame(table)[,which(colMeans(!is.na(data.frame(table))) >= call_rate)])

	if(ncol(table) > ncol(phenocovar)) {

	# Continue

	cases <- subset(table, table[[phenotype]] == 1)

	controls <- subset(table, table[[phenotype]] == 0)

	cases_cMAC <- sum(cases[,no_cols:ncol(cases)], na.rm = TRUE)

	cases_cMAF <- round(cases_cMAC/(sum(colSums(!is.na(cases[,no_cols:ncol(cases)])))*2), digits = 6)

	controls_cMAC <- sum(controls[,no_cols:ncol(controls)], na.rm = TRUE)

	controls_cMAF <- round(controls_cMAC/(sum(colSums(!is.na(controls[,no_cols:ncol(controls)])))*2), digits = 6)

	enrichment_ratio <- round(cases_cMAF/controls_cMAF, digits = 3)

	table$no_of_mutations <- rowSums(table[,no_cols:ncol(table)], na.rm = TRUE)

	no_of_carriers <- length(which(rowSums(table[,no_cols:ncol(table)], na.rm = TRUE) > 0))

	if(is.na(arg$covnames)){

	f = paste(paste("table$",phenotype, sep=""), "no_of_mutations",collapse="+table$",sep="~table$")

    } else if (!is.na(arg$covnames)){

	covlist = gsub(","," ", arg$covnames)

	covlist_prepared = unlist(strsplit(covlist,split=" "))

	f = paste(paste("table$",phenotype, sep=""), paste(c("no_of_mutations",covlist_prepared),collapse="+table$"),sep="~table$")

	}

	regression <- glm(as.formula(f), data = table, family=binomial(link="logit"))

	OR_noOfMutations <- round(as.numeric(exp(regression$coefficients[2])), digits = 3)

	OR_U95 <- round(exp(confint.default(regression, level = 0.95)[2,2]), digits = 3)

	OR_L95 <- round(exp(confint.default(regression, level = 0.95)[2,1]), digits = 3)

	eff_estimate <- round(regression$coefficients[2], digits = 3)

	eff_stdErr <- round(coef(summary(regression))[,"Std. Error"][2], digits = 3)

	P_regression <- as.numeric(coef(summary(regression))[,"Pr(>|z|)"][2])

	intermediate <- as.data.frame(cbind(SetID, cases_cMAC, cases_cMAF, controls_cMAC, controls_cMAF, no_of_carriers, enrichment_ratio, P_regression, OR_noOfMutations, OR_L95, OR_U95, eff_estimate, eff_stdErr))

	output_prepare <- rbind.data.frame(output_prepare, intermediate)

	} else {

	intermediate <- as.data.frame(cbind(SetID, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA))

	colnames(intermediate) <- c("SetID","cases_cMAC","cases_cMAF","controls_cMAC","controls_cMAF", "no_of_carriers", "enrichment_ratio", "P_regression", "OR_noOfMutations","OR_L95","OR_U95","eff_estimate","eff_stdErr")

	output_prepare <- rbind.data.frame(output_prepare, intermediate)

	} }

	output_ready <- output_prepare[-1,]

	output_ready_annotated <- left_join(output_ready,geneAnnotFile,by="SetID")

	merged <- left_join(results, output_ready_annotated, by="SetID")

	noNA_merged <- merged[!is.na(merged$P.value),]

	#Keep only the associations with at least carrier_cutoff in gene SetID

	noNA_merged$no_of_carriers <- as.numeric(noNA_merged$no_of_carriers)

	filtered_noNA_merged <- noNA_merged[noNA_merged$no_of_carriers >= arg$carrier_cutoff,]

	filtered_noNA_merged$P.value <- ifelse(filtered_noNA_merged$P.value >= 1, 1, filtered_noNA_merged$P.value)

	exome_wide_threshold <- 0.05/(nrow(filtered_noNA_merged))

	suggestive_threshold <- 1/(nrow(filtered_noNA_merged))

	filtered_noNA_merged$exome_wide_check <- ifelse(filtered_noNA_merged$P.value <= exome_wide_threshold, "exomeWideHit", ifelse(filtered_noNA_merged$P.value > exome_wide_threshold & filtered_noNA_merged$P.value <= suggestive_threshold , "suggestiveHit", NA))

	write.table(filtered_noNA_merged[order(P.value),],paste0("ANNOTATED.",resultname), col.names=TRUE, quote=F, row.names=F, sep="\t")

	#Gene Specific plots and tables

	selected_filtered_noNA_merged <- filtered_noNA_merged[filtered_noNA_merged$P.value <= arg$threshold | !is.na(filtered_noNA_merged$exome_wide_check),]

for (i in selected_filtered_noNA_merged$SetID){

	SetID <- i

	set_index <- ssd_info$SetInfo[ssd_info$SetInfo$SetID == i,]$SetIndex

	table <- cbind(phenocovar,as.data.frame(Get_Genotypes_SSD(ssd_info, set_index, is_ID = TRUE)))

	# Convert 9s in genotypes to NAs, and drop variants depending on the missing_cutoff within this subset to be in corcordant

	table[,no_cols:ncol(table)] <- table[,no_cols:ncol(table)] %>% mutate_all(~na_if(., 9))

	call_rate <- 1 - arg$missing_cutoff

	table <- data.table(data.frame(table)[,which(colMeans(!is.na(data.frame(table))) >= call_rate)])

	table$no_of_mutations <- rowSums(table[,no_cols:ncol(table)], na.rm = TRUE)

	write.table(table,paste0(plotAndTablePath,i,".annotTable.",resultname), col.names=TRUE, quote=F, row.names=F, sep="\t")

}

	#Now Manhattan and qqplot

	maxP = min(filtered_noNA_merged$P.value)
	b = -log10(maxP)
	ymax = b + 2

	nCHR <- length(unique(filtered_noNA_merged$chr))
	filtered_noNA_merged$StartPoscum <- 0
	s <- 0
	nbp <- c()
	for (i in unique(sort(filtered_noNA_merged$chr))){
	  nbp[i] <- max(filtered_noNA_merged[filtered_noNA_merged$chr == i,]$start)
	  filtered_noNA_merged[filtered_noNA_merged$chr == i,"StartPoscum"] <- filtered_noNA_merged[filtered_noNA_merged$chr == i,"start"] + s
	  s <- s + nbp[i]
	}

	axis.set <- filtered_noNA_merged %>%
	  group_by(chr) %>%
	  summarize(center = (max(StartPoscum) + min(StartPoscum)) / 2)

	is.odd <- function(x) x %% 2 != 0

	filtered_noNA_merged$chrcolor <- ifelse(is.odd(filtered_noNA_merged$chr), 'royalblue3', 'firebrick2')

	filtered_noNA_mergedLABEL <- filtered_noNA_merged[filtered_noNA_merged$P.value <= arg$threshold | !is.na(filtered_noNA_merged$exome_wide_check),]

	phenotypename <- arg$phenotype

	testtype <- arg$testtype

	png(paste0("manhattan.",resultname,".png"), height=10, width=15, res=600, units="in")
	q1 <- ggplot() + geom_point(data = filtered_noNA_merged, aes(x = StartPoscum, y = -log10(P.value)), color=filtered_noNA_merged$chrcolor) + scale_x_continuous(label = axis.set$chr, breaks = axis.set$center) + scale_y_continuous(limits = c(0,ymax),breaks = seq(0,ymax, by=1)) + geom_hline(yintercept=-log10(exome_wide_threshold), colour = "red", size = 0.5, linetype="dashed") + geom_hline(yintercept=-log10(suggestive_threshold), colour = "blue4", size = 0.25, linetype="dashed") + labs(x="Chromosomal Position",y=expression(-log[10](P))) + geom_text_repel(data=filtered_noNA_mergedLABEL, aes(x = StartPoscum, y = -log10(P.value), label = SetID), box.padding = 1, point.padding = 0.4, size = 4.5, segment.color = "black", min.segment.length = unit(0, 'lines'), nudge_y = .25, max.overlaps = getOption("ggrepel.max.overlaps", default = 1000)) + theme(panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"), axis.text.x = element_text(face="bold", size=15), axis.text.y = element_text(face="bold", size=15), axis.title.x = element_text(size=20), axis.title.y = element_text(size=20)) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "black")) + ggtitle(paste0(phenotypename," ",testtype," Manhattan Plot")) + theme(plot.title = element_text(hjust = 0.5, size=23))
	print(q1)
	dev.off()

	pvalues <- c(t(filtered_noNA_merged[,P.value]))
	qs <- qchisq(pvalues, df=1, lower.tail = TRUE)

	nonaqs <- qs[!is.na(qs)]

	nonapvalues <- pvalues[!is.na(pvalues)]

	lambda <- round(((median(nonaqs))/0.4549364),digits=3)
	png(paste0("qqplot.",resultname,".png"), width = 6, height = 6, res=600, units="in")
	qq(nonapvalues, main = paste0(phenotypename," ",testtype,"\nQQ Plot\nlambda = ",lambda))
	dev.off()

} else if (arg$variable == "continuous") {

	output_prepare <- as.data.frame(matrix(0, 1, 17))

	colnames(output_prepare) <- c("SetID","cMAC","cMAF", "medANDav_Pheno", "Carriers_AvPheno","nonCarriers_AvPheno", "AvDiff_Pheno", "Carriers_medianPheno","nonCarriers_medianPheno", "medianDiff_Pheno", "no_of_carriers", "P_regression", "BETA_noOfMutations","BETA_L95","BETA_U95","eff_estimate","eff_stdErr")

	for (i in results$SetID){

	SetID <- i

	set_index <- ssd_info$SetInfo[ssd_info$SetInfo$SetID == i,]$SetIndex

	table <- cbind(phenocovar,as.data.frame(Get_Genotypes_SSD(ssd_info, set_index, is_ID = TRUE)))

	# Convert 9s in genotypes to NAs, and drop variants depending on the missing_cutoff within this subset to be in corcordant

	table[,no_cols:ncol(table)] <- table[,no_cols:ncol(table)] %>% mutate_all(~na_if(., 9))

	call_rate <- 1 - arg$missing_cutoff

	table <- data.table(data.frame(table)[,which(colMeans(!is.na(data.frame(table))) >= call_rate)])

	if(ncol(table) > ncol(phenocovar)) {

	# Continue

	medANDav_Pheno <- paste0(round(median(table[[phenotype]]), digits=6),"_",round(mean(table[[phenotype]]), digits=6))

	carriers <- table[which(rowSums(table[,no_cols:ncol(table)], na.rm = TRUE) > 0)]

	nonCarriers <- table[!which(rowSums(table[,no_cols:ncol(table)], na.rm = TRUE) > 0)]

	cMAC <- sum(table[,no_cols:ncol(table)], na.rm = TRUE)

	cMAF <- round(cMAC/(sum(colSums(!is.na(table[,no_cols:ncol(table)])))*2), digits = 6)

	Carriers_AvPheno <- round(mean(carriers[[phenotype]]), digits = 3)

	nonCarriers_AvPheno <- round(mean(nonCarriers[[phenotype]]), digits = 3)

	AvDiff_Pheno <- round((mean(carriers[[phenotype]]) - mean(nonCarriers[[phenotype]])), digits = 3)

	Carriers_medianPheno <- round(median(carriers[[phenotype]]), digits = 3)

	nonCarriers_medianPheno <- round(median(nonCarriers[[phenotype]]), digits = 3)

	medianDiff_Pheno <- round((median(carriers[[phenotype]]) - median(nonCarriers[[phenotype]])), digits = 3)

	table$no_of_mutations <- rowSums(table[,no_cols:ncol(table)], na.rm = TRUE)

	no_of_carriers <- length(which(rowSums(table[,no_cols:ncol(table)], na.rm = TRUE) > 0))

	if(is.na(arg$covnames)){

	f = paste(paste("table$",phenotype, sep=""), "no_of_mutations",collapse="+table$",sep="~table$")

    } else if (!is.na(arg$covnames)){

	covlist = gsub(","," ", arg$covnames)

	covlist_prepared = unlist(strsplit(covlist,split=" "))

	}

	f = paste(paste("table$",phenotype, sep=""), paste(c("no_of_mutations",covlist_prepared),collapse="+table$"),sep="~table$")

	regression <- glm(as.formula(f), data = table, family=gaussian(link="identity"))

	BETA_noOfMutations <- round(as.numeric(regression$coefficients[2]), digits = 3)

	BETA_U95 <- round(confint.default(regression, level = 0.95)[2,2], digits = 3)

	BETA_L95 <- round(confint.default(regression, level = 0.95)[2,1], digits = 3)

	eff_estimate <- round(regression$coefficients[2], digits = 3)

	eff_stdErr <- round(coef(summary(regression))[,"Std. Error"][2], digits = 3)

	P_regression <- as.numeric(coef(summary(regression))[,"Pr(>|t|)"][2])

	intermediate <- as.data.frame(cbind(SetID, cMAC, cMAF, medANDav_Pheno, Carriers_AvPheno, nonCarriers_AvPheno, AvDiff_Pheno, Carriers_medianPheno, nonCarriers_medianPheno, medianDiff_Pheno, no_of_carriers, P_regression, BETA_noOfMutations, BETA_L95, BETA_U95, eff_estimate, eff_stdErr))

	output_prepare <- rbind.data.frame(output_prepare, intermediate)

	} else {

	intermediate <- as.data.frame(cbind(SetID, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA))

	colnames(intermediate) <- c("SetID","cMAC","cMAF", "medANDav_Pheno", "Carriers_AvPheno","nonCarriers_AvPheno", "AvDiff_Pheno", "Carriers_medianPheno","nonCarriers_medianPheno", "medianDiff_Pheno", "no_of_carriers", "P_regression", "BETA_noOfMutations","BETA_L95","BETA_U95","eff_estimate","eff_stdErr")

	output_prepare <- rbind.data.frame(output_prepare, intermediate)

	} }

	output_ready <- output_prepare[-1,]

	output_ready_annotated <- left_join(output_ready,geneAnnotFile,by="SetID")

	merged <- left_join(results, output_ready_annotated, by="SetID")

	noNA_merged <- merged[!is.na(merged$P.value),]

	#Keep only the associations with at least carrier_cutoff in gene SetID

	noNA_merged$no_of_carriers <- as.numeric(noNA_merged$no_of_carriers)

	filtered_noNA_merged <- noNA_merged[noNA_merged$no_of_carriers >= arg$carrier_cutoff,]

	filtered_noNA_merged$P.value <- ifelse(filtered_noNA_merged$P.value >= 1, 1, filtered_noNA_merged$P.value)

	exome_wide_threshold <- 0.05/(nrow(filtered_noNA_merged))

	suggestive_threshold <- 1/(nrow(filtered_noNA_merged))

	filtered_noNA_merged$exome_wide_check <- ifelse(filtered_noNA_merged$P.value <= exome_wide_threshold, "exomeWideHit", ifelse(filtered_noNA_merged$P.value > exome_wide_threshold & filtered_noNA_merged$P.value <= suggestive_threshold , "suggestiveHit", NA))

	write.table(filtered_noNA_merged[order(P.value),],paste0("ANNOTATED.",resultname), col.names=TRUE, quote=F, row.names=F, sep="\t")

	#Gene Specific plots and tables

	selected_filtered_noNA_merged <- filtered_noNA_merged[filtered_noNA_merged$P.value <= arg$threshold | !is.na(filtered_noNA_merged$exome_wide_check),]

for (i in selected_filtered_noNA_merged$SetID){

	SetID <- i

	set_index <- ssd_info$SetInfo[ssd_info$SetInfo$SetID == i,]$SetIndex

	table <- cbind(phenocovar,as.data.frame(Get_Genotypes_SSD(ssd_info, set_index, is_ID = TRUE)))

	# Convert 9s in genotypes to NAs, and drop variants depending on the missing_cutoff within this subset to be in corcordant

	table[,no_cols:ncol(table)] <- table[,no_cols:ncol(table)] %>% mutate_all(~na_if(., 9))

	call_rate <- 1 - arg$missing_cutoff

	table <- data.table(data.frame(table)[,which(colMeans(!is.na(data.frame(table))) >= call_rate)])

	table$no_of_mutations <- rowSums(table[,no_cols:ncol(table)], na.rm = TRUE)

	table$carrier <- ifelse(table$no_of_mutations > 0, "Carriers", "Non-carriers")

	write.table(table,paste0(plotAndTablePath,i,".annotTable.",resultname), col.names=TRUE, quote=F, row.names=F, sep="\t")

	cbPalette <- c("#D55E00", "#56B4E9", "#999999", "#E69F00", "#0072B2" , "#009E73", "#F0E442", "#CC79A7")

	min_1 <- round(min(table[[phenotype]]) - 1, digits=0)

	max_1 <- round(max(table[[phenotype]]) + 1, digits=0)

	phenotypename <- arg$phenotype

	testtype <- arg$testtype

	png(paste0(plotAndTablePath,i,".annotTable.",resultname,".png"), width = 6, height = 6, res=600, units="in")
	y1 <- ggplot(table, aes(x= table$carrier, y=table[[phenotype]], fill=factor(table$no_of_mutations))) + labs(fill="#Mutations\nCarried") + geom_violin(position=position_dodge(width=0.9)) + geom_beeswarm(dodge.width=0.9) + xlab(paste0(i, " ",testtype, " Mutations Carrier Status")) + ylab(paste0(phenotypename,"")) + scale_x_discrete(limits=c("Non-carriers", "Carriers")) + scale_y_continuous(breaks=seq(min_1, max_1, round((mean((table[[phenotype]])))/2))) + theme(legend.position = "right") + geom_boxplot(position=position_dodge(width=0.9), alpha=0, width = 0.25) + theme(axis.title.x = element_text(size=15), axis.title.y = element_text(size=15), legend.title = element_text(size=11)) + theme(axis.text.x = element_text(size=13), axis.text.y = element_text(size=13), legend.text = element_text(size=10)) + scale_fill_manual(values=cbPalette) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "black"))
	print(y1)
	dev.off()
}

	#Now Manhattan and qqplot

	maxP = min(filtered_noNA_merged$P.value)
	b = -log10(maxP)
	ymax = b + 2

	nCHR <- length(unique(filtered_noNA_merged$chr))
	filtered_noNA_merged$StartPoscum <- 0
	s <- 0
	nbp <- c()
	for (i in unique(sort(filtered_noNA_merged$chr))){
	  nbp[i] <- max(filtered_noNA_merged[filtered_noNA_merged$chr == i,]$start)
	  filtered_noNA_merged[filtered_noNA_merged$chr == i,"StartPoscum"] <- filtered_noNA_merged[filtered_noNA_merged$chr == i,"start"] + s
	  s <- s + nbp[i]
	}

	axis.set <- filtered_noNA_merged %>%
	  group_by(chr) %>%
	  summarize(center = (max(StartPoscum) + min(StartPoscum)) / 2)

	is.odd <- function(x) x %% 2 != 0

	filtered_noNA_merged$chrcolor <- ifelse(is.odd(filtered_noNA_merged$chr), 'royalblue3', 'firebrick2')

	filtered_noNA_mergedLABEL <- filtered_noNA_merged[filtered_noNA_merged$P.value <= arg$threshold | !is.na(filtered_noNA_merged$exome_wide_check),]

	phenotypename <- arg$phenotype

	testtype <- arg$testtype

	png(paste0("manhattan.",resultname,".png"), height=10, width=15, res=600, units="in")
	q1 <- ggplot() + geom_point(data = filtered_noNA_merged, aes(x = StartPoscum, y = -log10(P.value)), color=filtered_noNA_merged$chrcolor) + scale_x_continuous(label = axis.set$chr, breaks = axis.set$center) + scale_y_continuous(limits = c(0,ymax),breaks = seq(0,ymax, by=1)) + geom_hline(yintercept=-log10(exome_wide_threshold), colour = "red", size = 0.5, linetype="dashed") + geom_hline(yintercept=-log10(suggestive_threshold), colour = "blue4", size = 0.25, linetype="dashed") + labs(x="Chromosomal Position",y=expression(-log[10](P))) + geom_text_repel(data=filtered_noNA_mergedLABEL, aes(x = StartPoscum, y = -log10(P.value), label = SetID), box.padding = 1, point.padding = 0.4, size = 4.5, segment.color = "black", min.segment.length = unit(0, 'lines'), nudge_y = .25, max.overlaps = getOption("ggrepel.max.overlaps", default = 1000)) + theme(panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"), axis.text.x = element_text(face="bold", size=15), axis.text.y = element_text(face="bold", size=15), axis.title.x = element_text(size=20), axis.title.y = element_text(size=20)) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "black")) + ggtitle(paste0(phenotypename," ",testtype," Manhattan Plot")) + theme(plot.title = element_text(hjust = 0.5, size=23))
	print(q1)
	dev.off()

	pvalues <- c(t(filtered_noNA_merged[,P.value]))
	qs <- qchisq(pvalues, df=1, lower.tail = TRUE)

	nonaqs <- qs[!is.na(qs)]

	nonapvalues <- pvalues[!is.na(pvalues)]

	lambda <- round(((median(nonaqs))/0.4549364),digits=3)
	png(paste0("qqplot.",resultname,".png"), width = 6, height = 6, res=600, units="in")
	qq(nonapvalues, main = paste0(phenotypename," ",testtype,"\nQQ Plot\nlambda = ",lambda))
	dev.off()

}

} else if(arg$mode == "oneset") {

	if (arg$variable == "binary") {

	gene <- paste0(arg$gene,"")

	SetID <- gene

	output_prepare <- as.data.frame(matrix(0, 1, 13))

	colnames(output_prepare) <- c("SetID","cases_cMAC","cases_cMAF","controls_cMAC","controls_cMAF", "no_of_carriers", "enrichment_ratio", "P_regression", "OR_noOfMutations","OR_L95","OR_U95","eff_estimate","eff_stdErr")

	results_geneSpecific <- results[results$SetID == gene,]

	set_index <- ssd_info$SetInfo[ssd_info$SetInfo$SetID == gene,]$SetIndex

	table <- cbind(phenocovar,as.data.frame(Get_Genotypes_SSD(ssd_info, set_index, is_ID = TRUE)))

	# Convert 9s in genotypes to NAs, and drop variants depending on the missing_cutoff within this subset to be in corcordant

	table[,no_cols:ncol(table)] <- table[,no_cols:ncol(table)] %>% mutate_all(~na_if(., 9))

	call_rate <- 1 - arg$missing_cutoff

	table <- data.table(data.frame(table)[,which(colMeans(!is.na(data.frame(table))) >= call_rate)])

	if(ncol(table) > ncol(phenocovar)) {

	# Continue

	cases <- subset(table, table[[phenotype]] == 1)

	controls <- subset(table, table[[phenotype]] == 0)

	cases_cMAC <- sum(cases[,no_cols:ncol(cases)], na.rm = TRUE)

	cases_cMAF <- cases_cMAC/(sum(colSums(!is.na(cases[,no_cols:ncol(cases)])))*2)

	controls_cMAC <- sum(controls[,no_cols:ncol(controls)], na.rm = TRUE)

	controls_cMAF <- controls_cMAC/(sum(colSums(!is.na(controls[,no_cols:ncol(controls)])))*2)

	enrichment_ratio <- cases_cMAF/controls_cMAF

	table$no_of_mutations <- rowSums(table[,no_cols:ncol(table)], na.rm = TRUE)

	no_of_carriers <- length(which(rowSums(table[,no_cols:ncol(table)], na.rm = TRUE) > 0))

	if(is.na(arg$covnames)){

	f = paste(paste("table$",phenotype, sep=""), "no_of_mutations",collapse="+table$",sep="~table$")

    } else if (!is.na(arg$covnames)){

	covlist = gsub(","," ", arg$covnames)

	covlist_prepared = unlist(strsplit(covlist,split=" "))

	}
	
	f = paste(paste("table$",phenotype, sep=""), paste(c("no_of_mutations",covlist_prepared),collapse="+table$"),sep="~table$")

	regression <- glm(as.formula(f), data = table, family=binomial(link="logit"))

	OR_noOfMutations <- as.numeric(exp(regression$coefficients[2]))

	OR_U95 <- exp(confint.default(regression, level = 0.95)[2,2])

	OR_L95 <- exp(confint.default(regression, level = 0.95)[2,1])

	eff_estimate <- regression$coefficients[2]

	eff_stdErr <- coef(summary(regression))[,"Std. Error"][2]

	P_regression <- as.numeric(coef(summary(regression))[,"Pr(>|z|)"][2])

	intermediate <- as.data.frame(cbind(SetID, cases_cMAC, cases_cMAF, controls_cMAC, controls_cMAF, no_of_carriers, enrichment_ratio, P_regression, OR_noOfMutations, OR_L95, OR_U95, eff_estimate, eff_stdErr))

	output_prepare <- rbind.data.frame(output_prepare, intermediate)

	} else {

	intermediate <- as.data.frame(cbind(SetID, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA))

	colnames(intermediate) <- c("SetID","cases_cMAC","cases_cMAF","controls_cMAC","controls_cMAF", "no_of_carriers", "enrichment_ratio", "P_regression", "OR_noOfMutations","OR_L95","OR_U95","eff_estimate","eff_stdErr")

	output_prepare <- rbind.data.frame(output_prepare, intermediate)

	}

	output_ready <- output_prepare[-1,]

	output_ready_annotated <- left_join(output_ready,geneAnnotFile,by="SetID")

	merged <- left_join(results, output_ready_annotated, by="SetID")

	noNA_merged <- merged[!is.na(merged$P.value),]

	#Keep only the associations with at least carrier_cutoff in gene SetID

	noNA_merged$no_of_carriers <- as.numeric(noNA_merged$no_of_carriers)
	
	filtered_noNA_merged <- noNA_merged[noNA_merged$no_of_carriers >= arg$carrier_cutoff,]

	filtered_noNA_merged$P.value <- ifelse(filtered_noNA_merged$P.value >= 1, 1, filtered_noNA_merged$P.value)

	#exome_wide_threshold <- 0.05/(nrow(filtered_noNA_merged))

	#suggestive_threshold <- 1/(nrow(filtered_noNA_merged))

	filtered_noNA_merged$exome_wide_check <- "not_checked"

	write.table(table,paste0(plotAndTablePath,gene,".annotTable.",resultname), col.names=TRUE, quote=F, row.names=F, sep="\t")

	write.table(filtered_noNA_merged,paste0(gene,".oneSetResults.",resultname), col.names=TRUE, quote=F, row.names=F, sep="\t")

} else if (arg$variable == "continuous") {

	gene <- paste0(arg$gene,"")

	SetID <- gene

	output_prepare <- as.data.frame(matrix(0, 1, 17))

	colnames(output_prepare) <- c("SetID","cMAC","cMAF", "medANDav_Pheno", "Carriers_AvPheno","nonCarriers_AvPheno", "AvDiff_Pheno", "Carriers_medianPheno","nonCarriers_medianPheno", "medianDiff_Pheno", "no_of_carriers", "P_regression", "BETA_noOfMutations","BETA_L95","BETA_U95","eff_estimate","eff_stdErr")

	results_geneSpecific <- results[results$SetID == gene,]

	set_index <- ssd_info$SetInfo[ssd_info$SetInfo$SetID == gene,]$SetIndex

	table <- cbind(phenocovar,as.data.frame(Get_Genotypes_SSD(ssd_info, set_index, is_ID = TRUE)))

	# Convert 9s in genotypes to NAs, and drop variants depending on the missing_cutoff within this subset to be in corcordant

	table[,no_cols:ncol(table)] <- table[,no_cols:ncol(table)] %>% mutate_all(~na_if(., 9))

	call_rate <- 1 - arg$missing_cutoff

	table <- data.table(data.frame(table)[,which(colMeans(!is.na(data.frame(table))) >= call_rate)])
	
	if(ncol(table) > ncol(phenocovar)) {

	# Continue

	medANDav_Pheno <- paste0(round(median(table[[phenotype]]),digits=6),"_",round(mean(table[[phenotype]]),digits=6))

	carriers <- table[which(rowSums(table[,no_cols:ncol(table)], na.rm = TRUE) > 0)]

	nonCarriers <- table[!which(rowSums(table[,no_cols:ncol(table)], na.rm = TRUE) > 0)]

	cMAC <- sum(table[,no_cols:ncol(table)], na.rm = TRUE)

	cMAF <- cMAC/(sum(colSums(!is.na(table[,no_cols:ncol(table)])))*2)

	Carriers_AvPheno <- round(mean(carriers[[phenotype]]),digits=6)

	nonCarriers_AvPheno <- round(mean(nonCarriers[[phenotype]]),digits=6)

	AvDiff_Pheno <- round((mean(carriers[[phenotype]]) - mean(nonCarriers[[phenotype]])),digits=6)

	Carriers_medianPheno <- round(median(carriers[[phenotype]]),digits=6)

	nonCarriers_medianPheno <- round(median(nonCarriers[[phenotype]]),digits=6)

	medianDiff_Pheno <- round((median(carriers[[phenotype]]) - median(nonCarriers[[phenotype]])),digits=6)

	table$no_of_mutations <- rowSums(table[,no_cols:ncol(table)], na.rm = TRUE)

	no_of_carriers <- length(which(rowSums(table[,no_cols:ncol(table)], na.rm = TRUE) > 0))

	if(is.na(arg$covnames)){

	f = paste(paste("table$",phenotype, sep=""), "no_of_mutations",collapse="+table$",sep="~table$")

    } else if (!is.na(arg$covnames)){

	covlist = gsub(","," ", arg$covnames)

	covlist_prepared = unlist(strsplit(covlist,split=" "))

	}
	
	f = paste(paste("table$",phenotype, sep=""), paste(c("no_of_mutations",covlist_prepared),collapse="+table$"),sep="~table$")

	regression <- glm(as.formula(f), data = table, family=gaussian(link="identity"))

	BETA_noOfMutations <- as.numeric(regression$coefficients[2])

	BETA_U95 <- confint.default(regression, level = 0.95)[2,2]

	BETA_L95 <- confint.default(regression, level = 0.95)[2,1]

	eff_estimate <- regression$coefficients[2]

	eff_stdErr <- coef(summary(regression))[,"Std. Error"][2]

	P_regression <- as.numeric(coef(summary(regression))[,"Pr(>|t|)"][2])

	intermediate <- as.data.frame(cbind(SetID, cMAC, cMAF, medANDav_Pheno, Carriers_AvPheno, nonCarriers_AvPheno, AvDiff_Pheno, Carriers_medianPheno, nonCarriers_medianPheno, medianDiff_Pheno, no_of_carriers, P_regression, BETA_noOfMutations, BETA_L95, BETA_U95, eff_estimate, eff_stdErr))

	output_prepare <- rbind.data.frame(output_prepare, intermediate)

	} else {

	intermediate <- as.data.frame(cbind(SetID, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA))

	colnames(intermediate) <- c("SetID","cMAC","cMAF", "medANDav_Pheno", "Carriers_AvPheno","nonCarriers_AvPheno", "AvDiff_Pheno", "Carriers_medianPheno","nonCarriers_medianPheno", "medianDiff_Pheno", "no_of_carriers", "P_regression", "BETA_noOfMutations","BETA_L95","BETA_U95","eff_estimate","eff_stdErr")

	output_prepare <- rbind.data.frame(output_prepare, intermediate)

	}

	output_ready <- output_prepare[-1,]

	output_ready_annotated <- left_join(output_ready,geneAnnotFile,by="SetID")

	merged <- left_join(results, output_ready_annotated, by="SetID")

	noNA_merged <- merged[!is.na(merged$P.value),]

	#Keep only the associations with at least carrier_cutoff in gene SetID

	noNA_merged$no_of_carriers <- as.numeric(noNA_merged$no_of_carriers)
	
	filtered_noNA_merged <- noNA_merged[noNA_merged$no_of_carriers >= arg$carrier_cutoff,]

	filtered_noNA_merged$P.value <- ifelse(filtered_noNA_merged$P.value >= 1, 1, filtered_noNA_merged$P.value)

	#exome_wide_threshold <- 0.05/(nrow(filtered_noNA_merged))

	#suggestive_threshold <- 1/(nrow(filtered_noNA_merged))

	filtered_noNA_merged$exome_wide_check <- "not_checked"

	table$carrier <- ifelse(table$no_of_mutations > 0, "Carriers", "Non-carriers")

	write.table(table,paste0(plotAndTablePath,gene,".annotTable.",resultname), col.names=TRUE, quote=F, row.names=F, sep="\t")

	cbPalette <- c("#D55E00", "#56B4E9", "#999999", "#E69F00", "#0072B2" , "#009E73", "#F0E442", "#CC79A7")

	min_1 <- round(min(table[[phenotype]]) - 1, digits=0)

	max_1 <- round(max(table[[phenotype]]) + 1, digits=0)

	phenotypename <- arg$phenotype

	testtype <- arg$testtype

	png(paste0(plotAndTablePath,gene,".annotTable.",resultname,".png"), width = 6, height = 6, res=600, units="in")
	y1 <- ggplot(table, aes(x= table$carrier, y=table[[phenotype]], fill=factor(table$no_of_mutations))) + labs(fill="#Mutations\nCarried") + geom_violin(position=position_dodge(width=0.9)) + geom_beeswarm(dodge.width=0.9) + xlab(paste0(gene, " ",testtype, " Mutations Carrier Status")) + ylab(paste0(phenotypename,"")) + scale_x_discrete(limits=c("Non-carriers", "Carriers")) + scale_y_continuous(breaks=seq(min_1, max_1, round((mean((table[[phenotype]])))/2))) + theme(legend.position = "right") + geom_boxplot(position=position_dodge(width=0.9), alpha=0, width = 0.25) + theme(axis.title.x = element_text(size=15), axis.title.y = element_text(size=15), legend.title = element_text(size=11)) + theme(axis.text.x = element_text(size=13), axis.text.y = element_text(size=13), legend.text = element_text(size=10)) + scale_fill_manual(values=cbPalette) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "black"))
	print(y1)
	dev.off()

	write.table(filtered_noNA_merged,paste0(gene,".oneSetResults.",resultname), col.names=TRUE, quote=F, row.names=F, sep="\t")
}

}
